// houdini/build/runtime-cjs/lib/config.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var config_exports = {};
__export(config_exports, {
  computeID: () => computeID,
  defaultConfigValues: () => defaultConfigValues,
  getCurrentConfig: () => getCurrentConfig,
  getMockConfig: () => getMockConfig,
  keyFieldsForType: () => keyFieldsForType,
  localApiEndpoint: () => localApiEndpoint,
  localApiSessionKeys: () => localApiSessionKeys,
  setMockConfig: () => setMockConfig
});
module.exports = __toCommonJS(config_exports);
var import_config = __toESM(require("../imports/config"), 1);
var import_pluginConfig = __toESM(require("../imports/pluginConfig"), 1);
let mockConfig = null;
function getMockConfig() {
  return mockConfig;
}
function setMockConfig(config2) {
  mockConfig = config2;
}
function defaultConfigValues(file) {
  return {
    defaultKeys: ["id"],
    ...file,
    types: {
      Node: {
        keys: ["id"],
        resolve: {
          queryField: "node",
          arguments: (node) => ({ id: node.id })
        }
      },
      ...file.types
    }
  };
}
function keyFieldsForType(configFile, type) {
  const withDefault = defaultConfigValues(configFile);
  return withDefault.types?.[type]?.keys || withDefault.defaultKeys;
}
function computeID(configFile, type, data) {
  const fields = keyFieldsForType(configFile, type);
  let id = "";
  for (const field of fields) {
    id += data[field] + "__";
  }
  return id.slice(0, -2);
}
let _configFile = null;
function localApiEndpoint(configFile) {
  return configFile.router?.apiEndpoint ?? "/_api";
}
function localApiSessionKeys(configFile) {
  return configFile.router?.auth?.sessionKeys ?? [];
}
function getCurrentConfig() {
  const mockConfig2 = getMockConfig();
  if (mockConfig2) {
    return mockConfig2;
  }
  if (_configFile) {
    return _configFile;
  }
  let configFile = defaultConfigValues(import_config.default);
  for (const pluginConfig of import_pluginConfig.default) {
    configFile = pluginConfig(configFile);
  }
  _configFile = configFile;
  return configFile;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  computeID,
  defaultConfigValues,
  getCurrentConfig,
  getMockConfig,
  keyFieldsForType,
  localApiEndpoint,
  localApiSessionKeys,
  setMockConfig
});
