// houdini/build/runtime-cjs/cache/staleManager.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var staleManager_exports = {};
__export(staleManager_exports, {
  StaleManager: () => StaleManager
});
module.exports = __toCommonJS(staleManager_exports);
var import_lib = require("../lib");
class StaleManager {
  cache;
  fieldsTime = /* @__PURE__ */ new Map();
  constructor(cache) {
    this.cache = cache;
  }
  #initMapId = (id) => {
    if (!this.fieldsTime.get(id)) {
      this.fieldsTime.set(id, /* @__PURE__ */ new Map());
    }
  };
  getFieldTime(id, field) {
    return this.fieldsTime.get(id)?.get(field);
  }
  setFieldTimeToNow(id, field) {
    this.#initMapId(id);
    this.fieldsTime.get(id)?.set(field, new Date().valueOf());
  }
  markFieldStale(id, field) {
    this.#initMapId(id);
    this.fieldsTime.get(id)?.set(field, null);
  }
  markAllStale() {
    for (const [id, fieldMap] of this.fieldsTime.entries()) {
      for (const [field] of fieldMap.entries()) {
        this.markFieldStale(id, field);
      }
    }
  }
  markRecordStale(id) {
    const fieldsTimeOfType = this.fieldsTime.get(id);
    if (fieldsTimeOfType) {
      for (const [field] of fieldsTimeOfType.entries()) {
        this.markFieldStale(id, field);
      }
    }
  }
  markTypeStale(type) {
    for (const [id, fieldMap] of this.fieldsTime.entries()) {
      if (id.startsWith(`${type}:`)) {
        for (const [field] of fieldMap.entries()) {
          this.markFieldStale(id, field);
        }
      }
    }
  }
  markTypeFieldStale(type, field, when) {
    const key = (0, import_lib.computeKey)({ field, args: when });
    for (const [id, fieldMap] of this.fieldsTime.entries()) {
      if (id.startsWith(`${type}:`)) {
        for (const local_field of fieldMap.keys()) {
          if (local_field === key) {
            this.markFieldStale(id, field);
          }
        }
      }
    }
  }
  delete(id, field) {
    if (this.fieldsTime.has(id)) {
      this.fieldsTime.get(id)?.delete(field);
      if (this.fieldsTime.get(id)?.size === 0) {
        this.fieldsTime.delete(id);
      }
    }
  }
  reset() {
    this.fieldsTime.clear();
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  StaleManager
});
